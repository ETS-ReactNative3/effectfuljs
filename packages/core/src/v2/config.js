import config from "@effectful/transducers/v2/config";
Object.assign(config, {
  srcRoot: ".",
  filename: "file.js",
  rt: "@effectful/debugger/api",
  // information about the current scope
  evalContext: null,
  // uses the global variable as API namespace
  ns: null,
  // assumes the output is passed to some static bandler like webpack or parcel,
  // thus avoiding dynamic `require` injections
  staticBundler: true
});
export default config;
