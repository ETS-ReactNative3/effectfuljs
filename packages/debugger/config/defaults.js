const config = require("../config").default;
const fs = require("fs");
if (process.env.EFFECTFUL_DEBUGGER_URL)
  config.url = process.env.EFFECTFUL_DEBUGGER_URL;
config.timeTravel = isTrue(process.env.EFFECTFUL_DEBUGGER_TIME_TRAVEL);
config.timeTravelDisabled = isTrue(
  process.env.EFFECTFUL_DEBUGGER_TIME_TRAVEL_DISABLED
);
config.srcRoot = require("../state").normalizeDrive(
  fs.realpathSync(process.env.EFFECTFUL_DEBUGGER_SRC_ROOT || ".")
);
config.backend = process.env.EFFECTFUL_DEBUGGER_BACKEND || "vscode";
config.cache =
  isTrue(process.env.EFFECTFUL_DEBUGGER_CACHE) ||
  !isTrue(process.BABEL_DISABLE_CACHE);
config.zeroConfig = isTrue(process.env.EFFECTFUL_DEBUGGER_ZERO_CONFIG, true);
config.verbose = isNaN(process.env.EFFECTFUL_DEBUGGER_VERBOSE)
  ? isTrue(process.env.EFFECTFUL_DEBUGGER_VERBOSE)
    ? 1
    : 0
  : +process.env.EFFECTFUL_DEBUGGER_VERBOSE;
if (process.env.EFFECTFUL_DEBUGGER_PORT && !isNaN(process.env.EFFECTFUL_DEBUGGER_PORT))
  config.port = +process.env.EFFECTFUL_DEBUGGER_PORT;
else if (process.env.PORT && !isNaN(process.env.PORT)) {
  process.env.EFFECTFUL_DEBUGGER_PORT = process.env.PORT;
  config.port = +process.env.PORT;
}
if (process.env.EFFECTFUL_DEBUGGER_HOST)
  config.host = process.env.EFFECTFUL_DEBUGGER_HOST;
else if (process.env.HOST) {
  process.env.EFFECTFUL_DEBUGGER_HOST = process.env.HOST;
  config.host = process.env.HOST;
} else config.host = "0.0.0.0";

if (config.verbose) console.log(`DEBUGGER: config  ${JSON.stringify(config)}`);

config.open = isTrue(process.env.EFFECTFUL_DEBUGGER_OPEN);

config.instrument = isTrue(process.env.EFFECTFUL_DEBUGGER_INSTRUMENT, true);

config.instrumentDeps = isTrue(process.env.EFFECTFUL_DEBUGGER_INSTRUMENT_DEPS, true);

if (isTrue(process.env.EFFECTFUL_DEBUG_DEBUGGER)) config.debuggerDebug = true;

config.expEagerModuleExport = false;

config.builtInBackends = {
  vscode: true,
  standalone: true,
  standaloneTrace: true,
  react: true
};

module.exports = config;

function isTrue(str, defaultValue = false) {
  if (!str) return defaultValue;
  switch (str.toLowerCase()) {
    case "yes":
    case "y":
    case "1":
    case "t":
    case "true":
      return true;
    case "undefined":
    case "null":
    case "no":
    case "n":
    case "0":
    case "f":
    case "false":
      return false;
    default:
      throw new Error(`incorrect format for boolean: "${str}"`);
  }
}
