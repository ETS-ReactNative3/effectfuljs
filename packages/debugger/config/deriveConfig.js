const config = require("../config").default;
const fs = require("fs");
const path = require("path");
const mm = require("minimatch");
if (process.env.EFFECTFUL_DEBUGGER_URL)
  config.url = process.env.EFFECTFUL_DEBUGGER_URL;
config.timeTravel = isTrue(process.env.EFFECTFUL_DEBUGGER_TIME_TRAVEL);
config.timeTravelDisabled = isTrue(
  process.env.EFFECTFUL_DEBUGGER_TIME_TRAVEL_DISABLED
);

const { normalizePath } = require("../state");

config.srcRoot = fs.realpathSync(
  process.env.EFFECTFUL_DEBUGGER_SRC_ROOT || "."
);

const packageJSON = fs.realpathSync(
  require("find-up").sync("package.json", { cwd: config.srcRoot })
);

config.packageRoot = packageJSON ? path.dirname(packageJSON) : config.srcRoot;

config.runtimePackages =
  process.env.EFFECTFUL_DEBUGGER_RUNTIME_PACKAGES &&
  process.env.EFFECTFUL_DEBUGGER_RUNTIME_PACKAGES.length &&
  normalizePath(
    fs.realpathSync(process.env.EFFECTFUL_DEBUGGER_RUNTIME_PACKAGES)
  );

config.runtime =
  process.env.EFFECTFUL_DEBUGGER_RUNTIME || "@effectful/debugger";
config.cache =
  isTrue(process.env.EFFECTFUL_DEBUGGER_CACHE) ||
  !isTrue(process.BABEL_DISABLE_CACHE);
config.zeroConfig = isTrue(process.env.EFFECTFUL_DEBUGGER_ZERO_CONFIG, true);
config.verbose = isNaN(process.env.EFFECTFUL_DEBUGGER_VERBOSE)
  ? isTrue(process.env.EFFECTFUL_DEBUGGER_VERBOSE)
    ? 1
    : 0
  : +process.env.EFFECTFUL_DEBUGGER_VERBOSE;

if (
  process.env.EFFECTFUL_DEBUGGER_PORT &&
  !isNaN(process.env.EFFECTFUL_DEBUGGER_PORT)
)
  config.port = +process.env.EFFECTFUL_DEBUGGER_PORT;
else if (process.env.PORT && !isNaN(process.env.PORT)) {
  process.env.EFFECTFUL_DEBUGGER_PORT = process.env.PORT;
  config.port = +process.env.PORT;
}
if (process.env.EFFECTFUL_DEBUGGER_HOST)
  config.host = process.env.EFFECTFUL_DEBUGGER_HOST;
else if (process.env.HOST) {
  process.env.EFFECTFUL_DEBUGGER_HOST = process.env.HOST;
  config.host = process.env.HOST;
} else config.host = "0.0.0.0";

if (config.verbose) console.log(`DEBUGGER: config  ${JSON.stringify(config)}`);

config.open = isTrue(process.env.EFFECTFUL_DEBUGGER_OPEN);

config.instrument = isTrue(process.env.EFFECTFUL_DEBUGGER_INSTRUMENT, true);

config.instrumentDeps = isTrue(
  process.env.EFFECTFUL_DEBUGGER_INSTRUMENT_DEPS,
  true
);

config.blackbox = mm.makeRe(
  process.env.EFFECTFUL_DEBUGGER_BLACKBOX || "**/node_modules/**/*.?(m)js"
);
config.include = mm.makeRe(
  process.env.EFFECTFUL_DEBUGGER_INCLUDE || "**/*.[jte]s?(x)"
);
config.exclude = process.env.EFFECTFUL_DEBUGGER_EXCLUDE
  ? mm.makeRe(process.env.EFFECTFUL_DEBUGGER_EXCLUDE)
  : null;

if (isTrue(process.env.EFFECTFUL_DEBUG_DEBUGGER)) {
  config.debuggerDebug = true;
  Error.stackTraceLimit = Infinity;
}

config.expEagerModuleExport = false;

config.patchVM = isTrue(process.env.EFFECTFUL_DEBUGGER_PATCH_VM, true);

module.exports = config;

function isTrue(str, defaultValue = false) {
  if (!str) return defaultValue;
  switch (str.toLowerCase()) {
    case "yes":
    case "y":
    case "1":
    case "t":
    case "true":
      return true;
    case "undefined":
    case "null":
    case "no":
    case "n":
    case "0":
    case "f":
    case "false":
      return false;
    default:
      throw new Error(`incorrect format for boolean: "${str}"`);
  }
}
