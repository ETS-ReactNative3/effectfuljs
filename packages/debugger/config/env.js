/** assigning config parameters back to env variables, so they are propagated to children */
const path = require("path");

exports.fromArgv = function fromArgv(argv) {
  const env = {};
  if (argv.timeTravel) env.EFFECTFUL_DEBUGGER_TIME_TRAVEL = 1;
  if (argv.timeTravelDisabled) env.EFFECTFUL_DEBUGGER_TIME_TRAVEL_DISABLED = 1;
  if (argv.runtime) env.EFFECTFUL_DEBUGGER_RUNTIME = argv.runtime;
  if (argv.cache) env.EFFECTFUL_DEBUGGER_CACHE = 1;
  if (argv.open) env.EFFECTFUL_DEBUGGER_OPEN = 1;
  if (argv.url) env.EFFECTFUL_DEBUGGER_URL = argv.url;
  if (argv.srcRoot) env.EFFECTFUL_DEBUGGER_SRC_ROOT = argv.srcRoot;
  if (argv.port) env.EFFECTFUL_DEBUGGER_PORT = argv.port;
  if (argv.verbose) env.EFFECTFUL_DEBUGGER_VERBOSE = 1;
  if (argv.indexJs) env.EFFECTFUL_DEBUGGER_INDEX_JS = argv.indexJs;
  if (argv.instrument != null)
    env.EFFECTFUL_DEBUGGER_INSTRUMENT = argv.instrument;
  if (argv.instrumentDeps != null)
    env.EFFECTFUL_DEBUGGER_INSTRUMENT_DEPS = argv.instrumentDeps;
  if (argv.runtimePackages) {
    env.EFFECTFUL_DEBUGGER_RUNTIME_PACKAGES = argv.runtimePackages;
    if (!argv.defaultNodePath)
      env.NODE_PATH = process.env.NODE_PATH
        ? `${argv.runtimePackages}${path.delimiter}${process.env.NODE_PATH}`
        : argv.runtimePackages;
  }
  env.EFFECTFUL_DEBUGGER_ZERO_CONFIG = argv.zeroConfig ? 1 : 0;
  if (argv.htmlTemplate)
    env.EFFECTFUL_DEBUGGER_HTML_TEMPLATE = argv.htmlTemplate;
  if (argv.include) env.EFFECTFUL_DEBUGGER_INCLUDE = argv.include;
  env.EFFECTFUL_DEBUGGER_BLACKBOX = argv.blackbox;
  if (argv.exclude) env.EFFECTFUL_DEBUGGER_EXCLUDE = argv.exclude;
  env.EFFECTFUL_DEBUGGER_PATCH_VM = !!argv.patchVm;
  env.EFFECTFUL_DEBUGGER = 1;
  env.NODE_ENV = process.env.NODE_ENV || "development";
  return env;
};
