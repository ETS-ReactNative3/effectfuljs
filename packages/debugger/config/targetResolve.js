const config = require("./deriveConfig");

function targetResolve(request) {
  return require.resolve(request, { paths: [config.srcRoot] });
}

exports.targetResolve = targetResolve;

exports.targetRequire = function targetRequire(request) {
  return require(targetResolve(request));
};
