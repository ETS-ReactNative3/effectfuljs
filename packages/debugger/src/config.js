export default {
  /** replace standard runtime functions with their instrumented equivalents */
  replaceRT: true,
  /** enables traces collections */
  timeTravel: false,
  /** do whole object snapshot on each checkpoint  */
  objectSnapshots: true,
  /** do whole Map/Set snapshot on each checkpoint  */
  iterableSnapshots: false,
  /** experimental (soon to be removed) options for signaling effect with an exception */
  expTCExceptions: false
};
