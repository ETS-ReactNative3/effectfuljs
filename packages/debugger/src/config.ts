export default {
  /** replace standard runtime functions with their instrumented equivalents */
  patchRT: true,
  /** enables traces collections */
  timeTravel: false,
  /** don't start collecting traces immediately (awaits `EDBG.journal.enabled = true`) */
  timeTravelDisabled: false,
  /** tries to make everything serializable */
  persistState: true,
  /** Track DOM changes using MutationObserver */
  mutationObserver: true,
  /** Output debugging information to console (for some runtimes) */
  verbose: 0,
  /** URL to connect back to debugger adapter */
  url: "ws://localhost:20011",
  srcRoot: ".",
  /** module's hot swapping is enabled */
  hot: true,
  /** webpack-dev-server listening port */
  port: 12345,
  /** redirect `console` methods into the debug stream  */
  redirConsole: true,
  /** if `redirConsole` is specified output messages to a local console too */
  localConsole: true,
  expFunctionConstr: true,
  debuggerDebug: false,
  /** global variable NS for the debugger library */
  globalNS: "EDBG",
  keepObjectKeysOrder: true,
  keepContainersOrder: true,
  expNoAccessOverloading: true,
  expInlineNext: false,
  /** for timeTravel debugging don't exit node process on program termination */
  stopOnExit: false
};
