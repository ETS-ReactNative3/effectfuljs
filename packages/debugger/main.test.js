// otherwise jest won't work
require("./config").default.replaceRT = false;
const Debugger = require("./main");
const S = require("@effectful/serialization");
const path = require("path");

function trace() {
  let num = 0;
  let f;
  while ((f = Debugger.step()) === Debugger.token) {
    const x = Debugger.context;
    console.log(`Step #${num++}@${x.brk.kind}`, x.value);
    for (const j of x.stack) {
      const brk = j.brk;
      const pos = brk
        ? `${brk.line}:${brk.column}-${brk.endLine}:${brk.endColumn}`
        : "?";
      console.log(`  ${j.constructor.name}@${j.$meta.module.name}-${pos}`);
    }
  }
  return f;
}

test("modules tracing", function() {
  global.console = { log: jest.fn() };
  Debugger.unwrap(require("./__fixtures__/mod1"));
  const mod = trace();
  console.log("> M:", mod);
  Debugger.unwrap(require("./__fixtures__/mod1"));
  const mod1 = trace();
  console.log("> M1:", mod1);
  Debugger.unwrap(require("./__fixtures__/mod2"));
  const mod2 = trace();
  console.log("> M2:", mod2);

  mod.mapTest([1, 2, 3]);
  console.log("> map:Array", trace());
  mod.mapTest(new Int8Array([1, 2, 3]));
  console.log("> map:TypedArray", trace());

  mod.filterTest([1, 2, 3]);
  console.log("> filter:Array", trace());
  mod.filterTest(new Int8Array([1, 2, 3]));
  console.log("> filter:TypedArray", trace());

  mod.findIndexTest([1, 2, 3]);
  console.log("> findIndex:Array", trace());
  mod.findIndexTest(new Int8Array([1, 2, 3]));
  console.log("> findIndex:TypedArray", trace());

  mod.flatMapTest([1, 2, 3]);
  console.log("> flatMap:Array", trace());
  mod.flatMapTest(new Int8Array([1, 2, 3]));
  console.log("> flatMap:TypedArray", trace());

  mod.forEachTest([1, 2, 3]);
  console.log("> forEach:Array", trace());
  mod.forEachTest(new Int8Array([1, 2, 3]));
  console.log("> forEach:TypedArray", trace());
  mod.forEachTest(new Set([1, 2, 3]));
  console.log("> forEach:Set", trace());
  mod.forEachTest(new Map([[1, "a"], [2, "b"], [3, "c"]]));
  console.log("> forEach:Map", trace());

  mod.reduceTest(["a", "b", "c"]);
  console.log("> reduce:Array", trace());
  mod.reduceTest(new Int8Array([1, 2, 3]));
  console.log("> reduce:TypedArray", trace());

  mod.reduceRightTest(["a", "b", "c"]);
  console.log("> reduceRight:Array", trace());
  mod.reduceRightTest(new Int8Array([1, 2, 3]));
  console.log("> reduceRight:TypedArray", trace());

  mod.arraySortTest([2, 3, 1], ["b", "c", "a"]);
  console.log("> sort:Array", trace());
  mod.typedArraySortTest(new Int8Array([2, 3, 1]), new Int8Array([3, 2, 1]));
  console.log("> sort:TypedArray", trace());

  mod.someTest([1, 2, 3]);
  console.log("> some:Array", trace());
  mod.someTest(new Int8Array([1, 2, 3]));
  console.log("> some:TypedArray", trace());

  mod.everyTest([1, 2, 3]);
  console.log("> every:Array", trace());
  mod.everyTest(new Int8Array([1, 2, 3]));
  console.log("> every:TypedArray", trace());

  mod.runEval();
  console.log("> eval", trace());

  expect(console.log.mock.calls).toMatchSnapshot();
});

test("persistent state", function() {
  global.console = { log: jest.fn() };
  Debugger.unwrap(require("./__fixtures__/counters"));
  const mod = trace();
  S.regOpaqueObject(mod, "countersMod");
  mod.runCounters();
  const [counter1, s1] = trace();
  console.log(counter1, s1);
  Debugger.restore(s1);
  const [counter2, s2] = trace();
  console.log(counter2, s2);
  expect(s1).toEqual(s2);
  expect(console.log.mock.calls).toMatchSnapshot();
});
