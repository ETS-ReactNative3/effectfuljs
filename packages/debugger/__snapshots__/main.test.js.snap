// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tracing 1`] = `
Array [
  Array [
    "Step #0@statement",
    undefined,
  ],
  Array [
    "  f@/__fixtures__/mod1.js:1:0-1:29",
  ],
  Array [
    "Step #1@statement",
    undefined,
  ],
  Array [
    "  f@/__fixtures__/mod2.js:1:0-1:12",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:1:0-1:29",
  ],
  Array [
    "Step #2@return",
    Object {
      "__esModule": true,
      "f1": [Function],
      "f2": [Function],
    },
  ],
  Array [
    "  f@/__fixtures__/mod1.js:1:0-1:29",
  ],
  Array [
    "Step #3@statement",
    Object {
      "__esModule": true,
      "f1": [Function],
      "f2": [Function],
    },
  ],
  Array [
    "  f@/__fixtures__/mod1.js:3:0-3:10",
  ],
  Array [
    "Step #4@statement",
    Object {
      "__esModule": true,
      "f1": [Function],
      "f2": [Function],
    },
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #5@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:6:2-6:9",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #6@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #7@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:14:2-14:8",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #8@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:15:2-15:32",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "f1:1@1",
  ],
  Array [
    "Step #9@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #10@statement",
    undefined,
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:4:2-4:24",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #11@statement",
    undefined,
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #12@statement",
    undefined,
  ],
  Array [
    "  _B@/__fixtures__/mod2.js:9:2-9:13",
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #13@statement",
    undefined,
  ],
  Array [
    "  _B@/__fixtures__/mod2.js:10:2-10:26",
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #14@return",
    Object {
      "b": _B {
        "i": 100,
      },
      "cnt": 1,
    },
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #15@return",
    _A {
      "b": Object {
        "b": _B {
          "i": 100,
        },
        "cnt": 1,
      },
      "i": 11,
    },
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #16@statement",
    _A {
      "b": Object {
        "b": _B {
          "i": 100,
        },
        "cnt": 1,
      },
      "i": 11,
    },
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:17:2-17:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "A:",
    _A {
      "b": Object {
        "b": _B {
          "i": 100,
        },
        "cnt": 1,
      },
      "i": 11,
    },
  ],
  Array [
    "Step #17@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:18:2-18:13",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #18@return",
    1,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #19@statement",
    undefined,
  ],
  Array [
    "  _f2@/__fixtures__/mod2.js:22:2-22:8",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #20@statement",
    undefined,
  ],
  Array [
    "  _f2@/__fixtures__/mod2.js:23:2-23:32",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "f2:1@2",
  ],
  Array [
    "Step #21@return",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #22@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:8:2-8:9",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #23@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:9:2-9:31",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "i:2,j:3",
  ],
  Array [
    "Step #24@return",
    undefined,
  ],
  Array [
    "  f@/__fixtures__/mod1.js:12:0-12:6",
  ],
  Array [
    "Step #25@statement",
    undefined,
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #26@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:6:2-6:9",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #27@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #28@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:14:2-14:8",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #29@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:15:2-15:32",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "f1:1@3",
  ],
  Array [
    "Step #30@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #31@statement",
    undefined,
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:4:2-4:24",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #32@statement",
    undefined,
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #33@statement",
    undefined,
  ],
  Array [
    "  _B@/__fixtures__/mod2.js:9:2-9:13",
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #34@statement",
    undefined,
  ],
  Array [
    "  _B@/__fixtures__/mod2.js:10:2-10:26",
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #35@return",
    Object {
      "b": _B {
        "i": 100,
      },
      "cnt": 3,
    },
  ],
  Array [
    "  _A@/__fixtures__/mod2.js:5:2-5:22",
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #36@return",
    _A {
      "b": Object {
        "b": _B {
          "i": 100,
        },
        "cnt": 3,
      },
      "i": 33,
    },
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:16:2-16:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #37@statement",
    _A {
      "b": Object {
        "b": _B {
          "i": 100,
        },
        "cnt": 3,
      },
      "i": 33,
    },
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:17:2-17:23",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "A:",
    _A {
      "b": Object {
        "b": _B {
          "i": 100,
        },
        "cnt": 3,
      },
      "i": 33,
    },
  ],
  Array [
    "Step #38@statement",
    undefined,
  ],
  Array [
    "  _f1@/__fixtures__/mod2.js:18:2-18:13",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #39@return",
    3,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #40@statement",
    undefined,
  ],
  Array [
    "  _f2@/__fixtures__/mod2.js:22:2-22:8",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #41@statement",
    undefined,
  ],
  Array [
    "  _f2@/__fixtures__/mod2.js:23:2-23:32",
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "f2:3@4",
  ],
  Array [
    "Step #42@return",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:7:2-7:23",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #43@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:8:2-8:9",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #44@statement",
    undefined,
  ],
  Array [
    "  _run@/__fixtures__/mod1.js:9:2-9:31",
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "i:2,j:6",
  ],
  Array [
    "Step #45@return",
    undefined,
  ],
  Array [
    "  f@/__fixtures__/mod1.js:13:0-13:6",
  ],
  Array [
    "Step #46@return",
    Object {},
  ],
]
`;
